brontide+lnwire+wtwire+zpay32: migrate fuzz tests

babinchak opened the pull request:
## Change Description
Fixes #6679.
Migrates fuzz tests from using go-fuzz to using Go 1.18's new standard library fuzz testing.
All fuzz tests are moved into a file called fuzz_test.go for their respective package.
~~Makefile, scripts, docs also updated.~~ Moved this to #6862 

## Steps to Test
Go to each package (brontide, lnwire, watchtower/wtwire, zpay32) and run each fuzz test individually:
Example of running the brontide Fuzz_random_actone test on its own for 30s:
`cd brontide`
`go test -fuzz Fuzz_random_actone -fuzztime 30s`

PR #6862 will allow you to test in bulk:
run `make fuzz fuzztime=1s` to run every fuzz test for 1 second
See [docs/fuzz.md](https://github.com/babinchak/lnd/blob/fuzz_make/docs/fuzz.md) for further documentation

## Pull Request Checklist
### Testing
- [x] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included.
- [x] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [x] Any new logging statements use an appropriate subsystem and logging level.
- [x]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

babinchak commented:
make itest icase=test_multi-hop_htlc/committype=SCRIPT_ENFORCED_LEASE_zeroconf=true/local_force_close_on-chain_htlc_timeout
works for me locally (linux amd64), but not on travis.

Crypt-iQ reviewed (commented):
I think it would be better if each commit migrated a package's fuzz tests and also removed the original fuzz test in the `fuzz` package. That way it's more clear to see what's being changed (if anything) as a reviewer. I also think the comment changes can be done in the move commit rather than their own commit.

babinchak commented:
Removed original fuzz test package in each migration commit
Also removed references to go-fuzz in Makefile and scripts

lightninglabs-deploy commented:
@babinchak, remember to re-request review from reviewers when ready

babinchak commented:
In addition to the suggested changes, I also completely replaced make fuzz-build/fuzz-run with make fuzz. See docs/fuzz.md

Crypt-iQ commented on lnwire/fuzz_test.go:
```diff
@@ -0,0 +1,726 @@
+		// Prefix with MsgNodeAnnouncement.
+		data = prefixWithMsgType(data, MsgNodeAnnouncement)
+
+		// We have to do this here instead of in fuzz.Harness so that
```
nit: `fuzz.Harness` -> `harness`

Crypt-iQ commented on watchtower/wtwire/fuzz_test.go:
```diff
@@ -0,0 +1,183 @@
+
+	// Make sure byte array length (excluding 2 bytes for message type) is
+	// less than max payload size for the wire message. We check this because
+	// otherwise `go-fuzz` will keep creating inputs that crash on ReadMessage
```
nit: ref to `go-fuzz` can be removed

Crypt-iQ commented on watchtower/wtwire/fuzz_test.go:
```diff
@@ -0,0 +1,183 @@
+
+	msg, err := ReadMessage(r, 0)
+	if err != nil {
+		// go-fuzz generated []byte that cannot be represented as a
```
nit: go-fuzz ref can be rm

Crypt-iQ commented on lnwire/fuzz_test.go:
```diff
@@ -0,0 +1,726 @@
+	})
+}
+
+func Fuzz_query_short_chan_ids(f *testing.F) {
```
im sure the go people have their reasons, but imo the way they did this is dumb. you can't have fuzz tests with similar prefixes, so this one has to have a name change so that running `go test -run=^$ -fuzz=Fuzz_query_short_chan_ids` actually works

Crypt-iQ replied:
it seems like this is fine with `go test -run=^$ -fuzz=^Fuzz_reply_channel_range\$`

babinchak replied:
Will still change so no regex needed

babinchak replied:
Thought about putting _no_zlib at end of one
or moving zlib ahead in the prefix of the other
latter seems less awkward

Crypt-iQ reviewed (changes requested):
checkpointing the review here, am fuzzing with the new ones to see if i missed anything during review

Crypt-iQ reviewed (commented):
I think it would be easier for review if the makefile/doc changes were in a separate PR. Also can you squash the linting commits?

babinchak commented:
Fixed suggest changes
Not sure if I should linewrap areas like https://github.com/babinchak/lnd/blob/fuzz_upgrade/lnwire/fuzz_test.go#L149-L167
Kept that unwrapped, looks like other parts of the codebase use discretion when linewrapping.
Split linting commit and squashed

Crypt-iQ reviewed (approved):
lgtm 🗡️

lightninglabs-deploy commented:
@roasbeef: review reminder

lightninglabs-deploy commented:
@roasbeef: review reminder

lightninglabs-deploy commented:
@roasbeef: review reminder

Roasbeef reviewed (approved):
LGTM 🔗

Only ask here is that we move away from the current snake case naming for all the fuzz tests to the more conventional CamelCase naming (for Go).

guggero commented:
Needs a rebase.

morehouse commented:
Alternatively, I submitted a much smaller patch to quickfix the build 3 months ago. https://github.com/lightning/bolts/pull/851

Crypt-iQ commented:
@morehouse if you want to take over the pr from the current commit and do the name-change, we can get it merged

morehouse commented:
@Roasbeef @Crypt-iQ: Commandeered this PR with https://github.com/lightningnetwork/lnd/pull/7127.

guggero commented:
Replaced by https://github.com/lightningnetwork/lnd/pull/7127.
